### PR TITLE
Bug 1131526 - Running webapp-shared.js on node.js, r=ricky

### PR DESCRIPTION
**Pull request by cctuan**
https://bugzilla.mozilla.org/show_bug.cgi?id=1131526

### CODE REVIEW

**Comment by try-server-hook** [comment]
[cctuan George (cctuan)](https://github.com/cctuan) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=33b90f80c371)

**Comment by try-server-hook** [comment]
[cctuan George (cctuan)](https://github.com/cctuan) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=5e9db7e8e723)

File: build/utils-node.js [hunk 1]
```diff
@@ -228,15 +229,19 @@ module.exports = {
...
         // Skip error
         return;
       }
-      var file = this.getFile(filePath);
+      var file = self.getFile(filePath);
       if (!pattern || !(include ^ pattern.test(file.leafName))) {
-        files.push(this.getFile(file));
+        files.push(file);
       }
     }.bind(this));
```
Comment (RickyChien) [inline build/utils-node.js]: We already have ```bind(this)``` here, so that we can get rid of ```var self = this```.

File: build/utils-node.js [hunk 2]
```diff
@@ -443,7 +448,8 @@ module.exports = {
   },
 
   fileExists: function(path) {
-    return fs.accessSync(path);
+    return (fs.accessSync && fs.accessSync(path)) ||
+           (fs.existsSync && fs.existsSync(path));
   },
```
Comment (RickyChien) [inline build/utils-node.js]: Please remove ```fs.accessSync``` since it's buggy

File: build/webapp-shared.js [hunk 3]
```diff
@@ -142,19 +130,17 @@ WebappShared.prototype.copyPage = function(path) {
...
-  file.append('resources');
-
+  var paths = utils.joinPath(this.gaia.sharedFolder.path, 'resources');
   path.split('/').forEach(function(segment) {
-    file.append(segment);
-    if (utils.isSubjectToBranding(file.path)) {
-      file.append((this.config.OFFICIAL === '1') ? 'official' : 'unofficial');
+    paths = utils.joinPath(paths, segment);
+    if (utils.isSubjectToBranding(paths)) {
+      paths += (this.config.OFFICIAL === '1' ? '/official' : '/unofficial');
```
Comment (RickyChien) [inline build/webapp-shared.js]: Here might be a problem in windows with ```/```, we should use utils.joinPath

File: build/webapp-shared.js [hunk 4]
```diff
@@ -214,18 +200,19 @@ WebappShared.prototype.pushLocale = function(name) {
 };
 
 WebappShared.prototype.pushElements = function(path) {
-  var file = this.gaia.sharedFolder.clone();
-  file.append('elements');
+  var filePath = this.gaia.sharedFolder.path;
+  filePath += '/elements';
```
Comment (RickyChien) [inline build/webapp-shared.js]: use utils.joinPath()

File: build/webapp-shared.js [hunk 5]
```diff
@@ -214,18 +200,19 @@ WebappShared.prototype.pushLocale = function(name) {
...
 
   var elems = path.split('/');
 
   for (var i = 0; i < elems.length; i++) {
-    file.append(elems[i]);
+    filePath += '/' + elems[i];
```
Comment (RickyChien) [inline build/webapp-shared.js]: use utils.joinPath()

File: build/webapp-shared.js (hunk 4 above)
Comment (RickyChien) [inline build/webapp-shared.js]: In case of windows path issue, we should use utils.joinPath instead.

**Comment by try-server-hook** [comment]
[cctuan George (cctuan)](https://github.com/cctuan) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=343be91fa0f4)

**Pull request merged**